compiler/util/src/org/jetbrains/kotlin/utils/SmartHashMap.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.utils
+
+const val ARRAY_UNTIL_SIZE = 10
+
+/**
+ * [SmartHashMap] is a Map implementation that uses reference identity for keys.
+ * It uses 2 arrays to store keys & values until the number of entries stored is larger than 10.
+ * At that point it switches to using a HashMap and stays that way until [clear] is called.
+ * It does not auto convert back if the number of entries decreases through [remove].
+ */
+class SmartHashMap<K, V>() : MutableMap<K, V> {
+
+    private var keysArray: MutableList<K>? = arrayListOf()
```

Comment:
It will be faster if you just use `Array` directly of size `ARRAY_UNTIL_SIZE`. e.g., `private var keysArray: Array<K?>? = arrayOfNulls(ARRAY_UNTIL_SIZE)`. Use a simple `Int` property to keep track of the number of elements in the array. It will be tricky what to do about removals but perhaps you can use null to track empty slots. (keys cannot be null anyway)

Ditto for `valuesArray`

Reply:
that is a bit annoying as you then have to defrag the array manually. instead I used ArrayList<K>(ARRAY_UNTIL_SIZE) to preallocate the underlying array to 10, but still maintain the size semantics.